### PR TITLE
[REVIEW] Cugraph dgl block improvements

### PR DESCRIPTION
**Pull request by VibhuJawa**
This PR fixes: https://github.com/rapidsai/cugraph/issues/3784 and speeds up MFG creation by `3.5x` . 

Todo:
- [x] Add tests

Benchmarked on 6_462_743_488 edges with a batch size of `128` on a 1 V100:

Before PR Times:
```
1min 17s
```

After PR Times:
```
22 s
```

See link: https://gist.github.com/VibhuJawa/4852203f2e96de09d84d698af945682d

**Profiling:**

After PR: https://github.com/rapidsai/cugraph/pull/3810

image

image

Profile of splitting df into tensors : 

image

### CODE REVIEW

File: python/cugraph-dgl/cugraph_dgl/dataloading/utils/sampling_helpers.py [hunk 1]
```diff
@@ -136,82 +158,118 @@ def create_homogeneous_sampled_graphs_from_dataframe(
...
+                if seednodes_range is not None:
+                    max_dst_nodes = max(max_dst_nodes, seednodes_range)
+                size = (max_src_nodes, max_dst_nodes)
+                block = cugraph_dgl.nn.SparseGraph(
+                    src_ids=tensor_per_hop_d["sources"],
+                    dst_ids=tensor_per_hop_d["destinations"],
+                    size=size,
+                    formats=["csc"],
+                    reduce_memory=True,
+                )
```
[the remark this answers is not in the document]
Comment (tingyu66) [reply python/cugraph-dgl/cugraph_dgl/dataloading/utils/sampling_helpers.py:222]:
```suggestion
                max_src_nodes = tensor_per_hop_d["sources_range"]
                max_dst_nodes = tensor_per_hop_d["destinations_range"]
                if seednodes_range is not None:
                    max_dst_nodes = max(max_dst_nodes, seednodes_range)
                size = (max_src_nodes, max_dst_nodes)
                block = cugraph_dgl.nn.SparseGraph(
                    size,
                    src_ids=tensor_per_hop_d["sources"],
                    dst_ids=tensor_per_hop_d["destinations"],
                    formats=["csc"],
                    reduce_memory=True,
                )
```

`size` is now the first positional argument.

[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph-dgl/cugraph_dgl/dataloading/utils/sampling_helpers.py]: Resolved by https://github.com/rapidsai/cugraph/pull/3810/commits/1a7c87fab333717924bd9e32372a4a566a9dcf02

**Review by alexbarghi-nv — APPROVED** [review]
👍

**Comment by VibhuJawa** [comment]
@tingyu66 , Could you take another look please before we try to get this in.

File: python/cugraph-dgl/tests/test_utils.py [hunk 2]
```diff
@@ -140,3 +143,36 @@ def test_create_homogeneous_sampled_graphs_from_dataframe():
...
+        "sources": torch.as_tensor([0, 0, 1, 1], dtype=torch.int64, device="cuda"),
+        "destinations": torch.as_tensor([0, 0, 1, 2], dtype=torch.int64, device="cuda"),
+    }
+
+    seednodes_range = 10
+    sparse_graph = _create_homogeneous_cugraph_dgl_nn_sparse_graph(
+        tensor_d, seednodes_range
+    )
+    assert sparse_graph.num_sources() == 2
+    assert sparse_graph.num_destinations() == 10
```
Comment (tingyu66) [inline python/cugraph-dgl/tests/test_utils.py:177]:
```suggestion
    assert sparse_graph.num_src_nodes() == 2
    assert sparse_graph.num_dst_nodes() == 10
```

**Review by tingyu66 — CHANGES_REQUESTED** [review]
Please see my comments. Looks good otherwise.

File: python/cugraph-dgl/tests/test_utils.py (hunk 2 above)
Comment (VibhuJawa) [reply python/cugraph-dgl/tests/test_utils.py]: Thanks for checking this , resolved by https://github.com/rapidsai/cugraph/pull/3810/commits/f5e84961d9a3e7e0cc01b8d26611b02d4daea1ab

File: python/cugraph-dgl/tests/test_dataset.py [hunk 3]
```diff
@@ -112,7 +111,18 @@ def test_homogeneous_sampled_graphs_from_dataframe(seed_node):
...
+    else:
+        for dgl_block, cugraph_dgl_graph in zip(dgl_mfgs, cugraph_mfgs):
+            # Can not verify edge ids as they are not
+            # preserved in cugraph_dgl.nn.SparseGraph
+            assert dgl_block.num_src_nodes() == cugraph_dgl_graph.num_src_nodes()
+            assert dgl_block.num_dst_nodes() == cugraph_dgl_graph.num_dst_nodes()
+            dgl_offsets, dgl_indices, _ = dgl_block.adj_tensors("csc")
+            cugraph_offsets, cugraph_indices = cugraph_dgl_graph.csc()
+            torch.equal(dgl_offsets.to("cpu"), cugraph_offsets.to("cpu"))
+            torch.equal(dgl_indices.to("cpu"), cugraph_indices.to("cpu"))
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph-dgl/tests/test_dataset.py]: Yup, Done now. Thanks.

https://github.com/rapidsai/cugraph/pull/3810/commits/f5e84961d9a3e7e0cc01b8d26611b02d4daea1ab

**Pull request merged**